File: python/src/snake.py
import numpy as np

from .settings import SIZE, Direction, Arrow, Speed, SpeedMS
from .world import make_cubes

class Snake:
    def __init__(self):
        # Begin size 1 in the center
        self.segments = [np.array([SIZE / 2, SIZE / 2, SIZE / 2])]
        self.time_move = 0
        self.direction = Direction.FORWARD
        self.rotation = 0 # quarters of a circle clockwise

    def __len__(self):
        return len(self.segments)

    def update(self, deltatime=None, food=None):
        self.time_move += deltatime
        if len(self) <= Speed.SLOW.value and self.time_move > SpeedMS.SLOW.value:
            self.move(food)
            return

        if len(self) <= Speed.MEDIUM.value and self.time_move > SpeedMS.MEDIUM.value:
            self.move(food)
            return

        if self.time_move > SpeedMS.FAST.value:
            self.move(food)
            return

    def render(self):
        return make_cubes(self.segments)

    def turn(self, arrow):
        # TODO
        return None

    def move(self, food):
        n = self.segments[-1] + self.direction

        self.segments.append(n)
        if not np.array_equal(food, n):
            self.segments.pop()
        self.time_move = 0

    def isNextMovable(self):
        n = self.segments[-1] + self.direction

        # Map OOB
        if min(n) < 0 or max(n) < 32:
            return False
        
        # Self-collision
        return not self.isColliding(n)

    def isColliding(self, pos):
        for s in self.segments:
            if np.array_equal(s, pos):
                return True
        
        return False

File: project/src/snake.cpp
#include <vector>

#include <glm/glm.hpp>

#include "settings.hpp"
#include "time.hpp"

#include "snake.hpp"

const glm::vec3 Snake::top = glm::vec3(0, 1, 0);
const glm::vec3 Snake::bottom = glm::vec3(0, -1, 0);
const glm::vec3 Snake::right = glm::vec3(1, 0, 0);
const glm::vec3 Snake::left = glm::vec3(-1, 0, 0);
const glm::vec3 Snake::forward = glm::vec3(0, 0, 1);
const glm::vec3 Snake::back = glm::vec3(0, 0, -1);

Snake::Snake(Food &food) : food(food), segments(std::vector<glm::vec3>()), moveTimer(std::chrono::milliseconds(0)) {
  auto initial = glm::vec3(settings::size / 2, settings::size / 2, settings::size / 2);

  segments.push_back(initial - direction);
  segments.push_back(initial);

  food.generate(segments);
};

std::vector<glm::vec3> Snake::render() { return segments; }

void Snake::update() {
  moveTimer += Time::timeDelta;
  if (segments.size() <= settings::speed::slow) {
    if (moveTimer.count() > settings::speedMs::slow) {
      move();
    }
    return;
  }
  
  if (segments.size() <= settings::speed::medium) {
    if (moveTimer.count() > settings::speedMs::medium) {
      move();
    }
    return;
  }

  if (moveTimer.count() > settings::speedMs::fast) {
    move();
  }
}

void Snake::turn(Arrow arrow) {
  if (movePending) {
    return;
  }

  switch (arrow) {
  case Arrow::Top:
    if (direction != Snake::bottom) {
      direction = Snake::top;
    }
    break;
  case Arrow::Bottom:
    if (direction != Snake::top) {
      direction = Snake::bottom;
    }
    break;
  case Arrow::Right:
    if (direction != Snake::left) {
      direction = Snake::right;
    }
    break;
  case Arrow::Left:
    if (direction != Snake::right) {
      direction = Snake::left;
    }
    break;
  case Arrow::Forward:
    if (direction != Snake::back) {
      direction = Snake::forward;
    }
    break;
  case Arrow::Back:
    if (direction != Snake::forward) {
      direction = Snake::back;
    }
    break;
  }
}

void Snake::move() {
  auto next = segments.back() + direction;

  segments.push_back(next);
  if (!isColliding(food.position)) {
    segments.erase(segments.begin());
  } else {
    food.generate(segments);
  }

  movePending = false;
  moveTimer = std::chrono::milliseconds(0);
}

bool Snake::isCrashed() {
  if (segments.size() <= settings::speed::slow) {
    if (moveTimer.count() > settings::speedMs::slow) {
      return !isNextMovable();
    }
  }
  
  if (segments.size() <= settings::speed::medium) {
    if (moveTimer.count() > settings::speedMs::medium) {
      return !isNextMovable();
    }
  }

  if (moveTimer.count() > settings::speedMs::fast) {
    return !isNextMovable();
  }

  return false;
}

bool Snake::isNextMovable() {
  auto next = segments.back() + direction;

  // Map OOB
  if (next.x < 0 || next.y < 0 || next.z < 0 || next.x > settings::size || next.y > settings::size || next.z > settings::size) {
    return false;
  }

  // Self-collision
  return !isColliding(next);
}

bool Snake::isColliding(const glm::vec3 &pos) {
  for (auto s : segments) {
    if (s == pos) {
      return true;
    }
  }

  return false;
}
